### PR TITLE
Refactor Mentions Plugin V1 and V2

### PR DESCRIPTION
**Pull request by bkoatz**
Move existing "simple refactor" change into two separate mentions plugins

### CODE REVIEW

File: Hakawai/HKWMentionsPluginV2.m [hunk 1]
```diff
@@ -0,0 +1,1590 @@
...
+#import "HKWTextView.h"
+#import "HKWTextView+TextTransformation.h"
+#import "HKWTextView+Extras.h"
+#import "HKWTextView+Plugins.h"
+
+#import "HKWMentionsAttribute.h"
+
+#import "_HKWMentionsStartDetectionStateMachine.h"
+#import "_HKWMentionsCreationStateMachineV2.h"
+#import "_HKWMentionsCreationStateMachineV1.h"
```
Comment (li-jychen) [inline Hakawai/HKWMentionsPluginV2.m:27]: why does it need v1.h?

File: Hakawai/HKWMentionsPluginV2.m [hunk 2]
```diff
@@ -0,0 +1,1590 @@
...
+ */
+@property (nonatomic) BOOL nextSelectionChangeShouldBeIgnored;
+
+/*!
+ Indicates that the next call to \c textView:shouldChangeTextInRange:replacementText: should be ignored. This property
+ is a fix for an issue where, when a suggestion-based keyboard suggestion is deleted (e.g. for the Chinese keyboard),
+ the delegate method is fired requesting the suggestion be deleted, and then immediately again with a spurious
+ insertion.
+ */
+@property (nonatomic) BOOL nextInsertionShouldBeIgnored;
```
Comment (li-jychen) [inline Hakawai/HKWMentionsPluginV2.m:66]: looks like they are not actually being used, can we remove?

File: Hakawai/HKWMentionsPluginV2.m [hunk 3]
```diff
@@ -0,0 +1,1590 @@
...
+ this should not be used to determine whether or not the callback is synchronous.)
+ */
+@property (nonatomic) unichar characterForAdvanceStateForCharacterInsertion;
+
+// Properties for resuming mentions creation
+@property (nonatomic) BOOL shouldResumeMentionsCreation;
+@property (nonatomic) unichar resumeMentionsControlCharacter;   // 0 if implicit mention
+@property (nonatomic) NSUInteger resumeMentionsPriorTextLength;
+@property (nonatomic) NSUInteger resumeMentionsPriorPosition;
+@property (nonatomic) NSString *resumeMentionsPriorString;
```
Comment (li-jychen) [inline Hakawai/HKWMentionsPluginV2.m:124]: can you double check if all those properties are actually being used?

File: Hakawai/HKWMentionsPluginV2.m [hunk 4]
```diff
@@ -0,0 +1,1590 @@
...
+
+/*!
+ Returns whether a string is a single control character
+ */
+- (BOOL)isStringControlCharacter:(NSString *)text {
+    return text.length == 1 && [self.controlCharacterSet characterIsMember:[text characterAtIndex:0]];
+}
+
+// TODO: Remove text view from call
+// JIRA: XXXX
```
Comment (li-jychen) [inline Hakawai/HKWMentionsPluginV2.m:1005]: what does this comment mean?

File: Hakawai/Mentions/HKWMentionsPlugin.h [hunk 5]
```diff
@@ -17,11 +17,34 @@
 #import "HKWChooserViewProtocol.h"
 #import "HKWMentionsDefaultChooserViewDelegate.h"
 
+static NSString* _Nonnull const HKWMentionAttributeName = @"HKWMentionAttributeName";
+
+// Don't confuse this with the public 'HKWMentionsPluginState', which exposes fewer implementation details.
+typedef NS_ENUM(NSInteger, HKWMentionsState) {
```
Comment (li-jychen) [inline Hakawai/Mentions/HKWMentionsPlugin.h:23]: so in simple refactor. this is the only state we will be managing?

File: Hakawai/Mentions/HKWMentionsPlugin.h [hunk 6]
```diff
@@ -142,8 +165,17 @@ typedef NS_ENUM(NSInteger, HKWMentionsPluginState) {
...
+ This is a temprorary protocol that V1 and V2 version of HKWMentionsPlugin will conform to so we can toggle between two versions easily.
+ Once V2 is fully ramped and tested, we will remove V1 version and this protocol.
+ */
+@protocol HKWMentionsPlugin <HKWDirectControlFlowPluginProtocol>
 
+@property (nonatomic, nonnull, strong) NSCharacterSet *controlCharacterSet;
+@property (nonatomic) NSInteger implicitSearchLength;
+@property (nonatomic, readonly) BOOL implicitMentionsEnabled;
+
+@property (nonatomic) BOOL shouldEnableUndoUponUnregistration;
```
Comment (li-jychen) [inline Hakawai/Mentions/HKWMentionsPlugin.h:178]: i see they were originally from `_mentionsPlugin.h` file, is the plan to move them back after we clean up ?

File: Hakawai/HKWMentionsPluginV2.m (hunk 2 above)
Comment (bkoatz) [reply Hakawai/HKWMentionsPluginV2.m:66]: yes removed a lot more

File: Hakawai/HKWMentionsPluginV2.m (hunk 3 above)
Comment (bkoatz) [reply Hakawai/HKWMentionsPluginV2.m:124]: yes removed

File: Hakawai/HKWMentionsPluginV2.m (hunk 4 above)
Comment (bkoatz) [reply Hakawai/HKWMentionsPluginV2.m:1005]: added jira. And it's that we don't need the textView arg in the call after v2/v1 split is cleaned up

File: Hakawai/Mentions/HKWMentionsPlugin.h (hunk 5 above)
Comment (bkoatz) [reply Hakawai/Mentions/HKWMentionsPlugin.h:23]: removed this. we don't need it

File: Hakawai/Mentions/HKWMentionsPlugin.h (hunk 6 above)
Comment (bkoatz) [reply Hakawai/Mentions/HKWMentionsPlugin.h:178]: yes. added comment

File: Hakawai/HKWMentionsPluginV2.m [hunk 7]
```diff
@@ -0,0 +1,1372 @@
...
+ */
+- (unichar)getExplicitSearchControlCharacter {
+    return self.creationStateMachine.explicitSearchControlCharacter;
+}
+
+// TODO: Remove this call
+// JIRA: POST-14031
+- (void)textViewDidProgrammaticallyUpdate:(UITextView * _Null_unspecified)textView {
+    return;
+}
```
Comment (li-jychen) [inline Hakawai/HKWMentionsPluginV2.m:1346]: does it not cause any regression?

Comment (bkoatz) [reply Hakawai/HKWMentionsPluginV2.m:1346]: no. cursor movement should handle all this

**Pull request merged**